aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,336 @@
+    Bytes(BytesPartReader),
+    Fs(PathBodyPartReader),
+}
+
+impl ReadPart for PartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        match self {
+            PartReader::Bytes(bytes) => bytes.next_part().await,
+            PartReader::Fs(path_body) => path_body.next_part().await,
+        }
+    }
+}
+
+/// Data for a single part
+pub(crate) struct PartData {
+    pub(crate) data: Bytes,
+    pub(crate) part_number: u64,
+}
+
+/// The `ReadPart` trait allows for reading data from an `InputStream` and packaging the raw
+/// data into `PartData` which carries additional metadata needed for uploading a part.
+pub(crate) trait ReadPart {
+    /// Request the next "part" of data.
+    ///
+    /// When there is no more data readers should return `Ok(None)`.
+    /// NOTE: Implementations are allowed to return data in any order and consumers are
+    /// expected to order data by the part number.
+    fn next_part(
+        &self,
+    ) -> impl std::future::Future<Output = Result<Option<PartData>, Error>> + Send;
+}
+
+#[derive(Debug)]
+struct PartReaderState {
+    // current start offset
+    offset: u64,
+    // current part number
+    part_number: u64,
+    // total number of bytes remaining to be read
+    remaining: u64,
+}
+
+impl PartReaderState {
+    /// Create a new `PartReaderState`
+    fn new(content_length: u64) -> Self {
+        Self {
+            offset: 0,
+            part_number: 1,
+            remaining: content_length,
+        }
+    }
+
+    /// Set the initial offset to start reading from
+    fn with_offset(self, offset: u64) -> Self {
+        Self { offset, ..self }
+    }
+}
+
+/// [ReadPart] implementation for in-memory input streams.
+#[derive(Debug)]
+struct BytesPartReader {
+    buf: Bytes,
+    part_size: usize,
+    state: Mutex<PartReaderState>,
+}
+
+impl BytesPartReader {
+    fn new(buf: Bytes, part_size: usize) -> Self {
+        let content_length = buf.remaining() as u64;
+        Self {
+            buf,
+            part_size,
+            state: Mutex::new(PartReaderState::new(content_length)),
+        }
+    }
+}
+
+impl ReadPart for BytesPartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        let mut state = self.state.lock().expect("lock valid");
+        if state.remaining == 0 {
+            return Ok(None);
+        }
+
+        let start = state.offset as usize;
+        let end = cmp::min(start + self.part_size, self.buf.len());
+        let data = self.buf.slice(start..end);
+        let part_number = state.part_number;
+        state.part_number += 1;
+        state.offset += data.len() as u64;
+        state.remaining -= data.len() as u64;
+        let part = PartData { data, part_number };
+        Ok(Some(part))
+    }
+}
+
+/// [ReadPart] implementation for path based input streams
+#[derive(Debug)]
+struct PathBodyPartReader {
+    body: PathBody,
+    part_size: usize,
+    state: Mutex<PartReaderState>,
```

Comment:
For ease of reading struct defs, would you mind adding a comment saying what variety of Mutex it is? That way I don't have to scroll back to the `use` statements.
```suggestion
    state: Mutex<PartReaderState>, // std Mutex
```